### PR TITLE
[ci skip] add class level documentation to ActiveModel::Type::Boolean

### PR DESCRIPTION
**Pull request by taboularasa**
### Summary

adds documentation of the behaviors of type coercion at the class level to ActiveModel::Type::Boolean
### Other Information

@sgrif per our conversation last week, let me know if this is what you had in mind.

### CODE REVIEW

File: activemodel/lib/active_model/type/boolean.rb [hunk 1]
```diff
@@ -1,9 +1,24 @@
...
+      # == Active \Model \Type \Boolean
+      #
+      # A class representing +boolean+ with functionality for coercion to and from
+      # a database or user input.
+      #
+      # === Coercion
+      # Values set from user or database input will first be coerced into a
+      # appropriate ruby type.
+      #
+      # - Values specified in +FALSE_VALUES+ will be coerced to +false+
```
Comment (sgrif) [inline activemodel/lib/active_model/type/boolean.rb]: We should mention the most common ones explicitly, `false, "0", 0, and 'f'`.

File: activemodel/lib/active_model/type/boolean.rb [hunk 2]
```diff
@@ -1,9 +1,24 @@
...
+      # - An empty string will be coerced to +nil+
+      # - +nil+ will not be coerced
```
Comment (sgrif) [inline activemodel/lib/active_model/type/boolean.rb]: I think this can be assumed

File: activemodel/lib/active_model/type/boolean.rb [hunk 3]
```diff
@@ -1,9 +1,24 @@
 module ActiveModel
   module Type
-    class Boolean < Value # :nodoc:
+    class Boolean < Value
+
+      # == Active \Model \Type \Boolean
```
Comment (rafaelfranca) [inline activemodel/lib/active_model/type/boolean.rb]: This should be above the `class` definition.

**Comment by taboularasa** [comment]
@sgrif @rafaelfranca all feedback addressed

File: activemodel/lib/active_model/type/boolean.rb [hunk 4]
```diff
@@ -1,9 +1,22 @@
...
+    #
+    # A class representing +boolean+ with functionality for coercion to and from
+    # a database or user input.
+    #
+    # === Coercion
+    # Values set from user or database input will first be coerced into a
+    # appropriate ruby type.
+    #
+    # - Values specified in +FALSE_VALUES+ will be coerced to +false+.
+    #   - e.g. "false", "f" , "0", and +0+ will all be coerced to +false+
```
Comment (sgrif) [inline activemodel/lib/active_model/type/boolean.rb]: I'd prefer this to be a bit more human. Something like:

`false`, `"f"`, `"0"`, `0`, or any other value in `FALSE_VALUES` will be coerced to `false`

File: activemodel/lib/active_model/type/boolean.rb [hunk 5]
```diff
@@ -1,9 +1,22 @@
 module ActiveModel
   module Type
-    class Boolean < Value # :nodoc:
+    # == Active \Model \Type \Boolean
+    #
+    # A class representing +boolean+ with functionality for coercion to and from
+    # a database or user input.
```
Comment (sgrif) [inline activemodel/lib/active_model/type/boolean.rb]: This sentence feels way to clinical for lack of a better word, but I don't have a specific suggestion to improve it.

File: activemodel/lib/active_model/type/boolean.rb [hunk 6]
```diff
@@ -1,9 +1,22 @@
...
+    #
+    # === Coercion
+    # Values set from user or database input will first be coerced into a
```
Comment (sgrif) [inline activemodel/lib/active_model/type/boolean.rb]: From a public API point of view, how database input is coerced is entirely irrelevant. We only need to focus on what happens during assignment from rubyland

File: activemodel/lib/active_model/type/boolean.rb [hunk 7]
```diff
@@ -1,9 +1,22 @@
...
+    # appropriate ruby type.
+    #
+    # - Values specified in +FALSE_VALUES+ will be coerced to +false+.
+    #   - e.g. "false", "f" , "0", and +0+ will all be coerced to +false+
+    # - An empty string will be coerced to +nil+
```
Comment (sgrif) [inline activemodel/lib/active_model/type/boolean.rb]: I would probably phrase this as `Empty strings are coerced to nil`

File: activemodel/lib/active_model/type/boolean.rb [hunk 8]
```diff
@@ -1,9 +1,22 @@
...
+    # - All other values will be coerced to +true+
```
Comment (sgrif) [inline activemodel/lib/active_model/type/boolean.rb]: This description is correct, but I feel like we should mention somewhere that this attempts to roughly map to Ruby's boolean semantics (not necessarily on this line, this is just the line that made me think of it)

**Comment by sgrif** [comment]
TIL I have more opinions than I realized about how to document type coercion to booleans.

**Comment by sgrif** [comment]
Thanks for working on this @taboularasa :heart:

**Comment by taboularasa** [comment]
@sgrif thanks for your feedback and please keep it coming, I would like to get a good sense for this before I work on documenting the rest of the `ActiveModel::Type` namespace

**Comment by taboularasa** [comment]
@sgrif ok all feedback addressed again, please take another look when you get a chance

**Comment by sgrif** [comment]
I'm about to head to bed, and I'm in Europe for a few weeks as of tomorrow, but I will try to take a look soon (but it won't be until Saturday at the earliest)

**Comment by taboularasa** [comment]
Np, take your time

File: activemodel/lib/active_model/type/boolean.rb [hunk 9]
```diff
@@ -1,9 +1,20 @@
 module ActiveModel
   module Type
-    class Boolean < Value # :nodoc:
+    # == Active \Model \Type \Boolean
+    #
+    # A class that behaves like a boolean type, including rules for coercion of user input.
+    #
+    # === Coercion
+    # Values set from user input will first be coerced into the appropriate ruby type.
+    # Coercion behaivor is roughly mapped to Ruby's boolean semantics.
```
Comment (vijaydev) [inline activemodel/lib/active_model/type/boolean.rb]: s/behaivor/behavior

**Comment by sgrif** [comment]
LGTM other than the typo.

**Comment by taboularasa** [comment]
typo fixed and rebased

**Pull request merged**